### PR TITLE
refactor(controllers): Refactored how Controllers interact w/ three

### PR DESCRIPTION
**Pull request by saitonakamura**
This a big refactor of controller and how they interact with three counterparts. The flow of events was very confusing, to say the least. This PR makes this flow much direct and does some other notable changes.

### ControllerModel abstraction is removed

It's responsibilities are moved to `Controllers`, mainly to https://github.com/pmndrs/react-xr/pull/294/files#diff-f17a5c6c1d801f88dd02bbbcaacc32ee99c5630535d1db5f3d5527ddd84b89eaR85-R103

And `Controllers` now directly render `XRControllerModel`

#### So the flow now is

1. `XR` creates a pair of `XRController` and subscribes to `connected`/`disconnected` events from three.js
2. `XRController` call `xr.getControler` and others in constructor therefore creating a `WebXRController` instances in three.js. Both `WebXRController` and it's `XRController` counterparts will now be present for the lifetime of an app
3. three.js `WebXRManager` react to `inputsourceschange` and populates it's state and dispatches `connected` event
4. `connected` event is handled by `XR` and it populates `state.controller` with corresponding `XRController` 
5. `handleControllerModel` react to `ref` and calls `modelFactory.initializeControllerModel` which does all the 3d/profiles heavy lifting as previous

Previously, the 5th step was actually another multi-step flow which involved sending fake events and such

### XRControllerModel moved to a separate file from XRControllersModelFactory

No notable changes inside it

### InputSource on XRController can now be null

Now inputSource will become null if controller is disconnected, no real reason to hold on to outdated inputSource. This resulted in a bunch of null checks here and there, but nothing substantial

### Whole bunch of unit tests added

We're still far from good coverage, but this is a big step forward

### CODE REVIEW

**Comment by saitonakamura** [comment]
@CodyJasonBennett I'd really appreciate any feedback on this

**Comment by CodyJasonBennett** [comment]
I'm happy with the premise. Will give this a deeper review later today.

File: src/Controllers.tsx [hunk 1]
```diff
@@ -111,20 +80,33 @@ export function Controllers({ rayMaterial = {}, hideRaysOnBlur = false }: Contro
...
     }
-  }, [controllers])
+  }, [])
 
   return (
     <>
       {controllers.map((target, i) => (
         <React.Fragment key={i}>
-          {createPortal(<controllerModel args={[target]} />, target.grip)}
+          {createPortal(<xRControllerModel ref={(r) => handleControllerModel(r, target)} args={[]} />, target.grip)}
```
Comment (CodyJasonBennett) [inline src/Controllers.tsx:109]: Need to move `target` to `useCallback` deps so we can get a stable reference here between successive renders.

**Review by CodyJasonBennett — CHANGES_REQUESTED** [review]

File: src/XRController.ts [hunk 2]
```diff
@@ -1,12 +1,16 @@
...
 
+/** Counterpart of WebXRController from three js
+ * in a sense that it's long living */
 export class XRController extends THREE.Group {
   readonly index: number
   readonly controller: THREE.XRTargetRaySpace
   readonly grip: THREE.XRGripSpace
   readonly hand: THREE.XRHandSpace
-  public inputSource!: XRInputSource
+  public inputSource: XRInputSource | null = null
```
Comment (CodyJasonBennett) [inline src/XRController.ts:12]: I'm not considering this breaking since it was wrongly implicit before.

File: src/Controllers.tsx (hunk 1 above)
Comment (saitonakamura) [reply src/Controllers.tsx]: Ooh, that's a good catch, thank you. I wrote a test and fixed it by moving it to ControllerModel component that will accept target now in 3a5039420f19b66ea1b1074ad88d9eb11739beb5

File: src/testUtils/testUtilsThree.tsx [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+  if (typeof window !== 'undefined') {
+    return window
+  }
+  /* istanbul ignore next */
+  if (typeof global !== 'undefined') {
+    return global
+  }
+  /* istanbul ignore next */
+  throw new Error('unable to locate global object')
+}
```
Comment (CodyJasonBennett) [inline src/testUtils/testUtilsThree.tsx:25]: In order for `act` to work, the scheduler has to be mocked to `scheduler/unstable_mock`. Perhaps `IS_REACT_ACT_ENVIRONMENT` etc. can be moved to a setup file?

Comment (saitonakamura) [reply src/testUtils/testUtilsThree.tsx:25]: > Perhaps IS_REACT_ACT_ENVIRONMENT etc. can be moved to a setup file?

I honestly don't know how it's supposed to be handled. In react testing library they wrap the act itself, set `IS_REACT_ACT_ENVIRONMENT` to true before executing callback in act and set `IS_REACT_ACT_ENVIRONMENT` back to previous value when act is completed. I'm not sure why they do it, but this is the reason why I thought that setting it to true across all test via setupFile.

Comment (saitonakamura) [reply src/testUtils/testUtilsThree.tsx:25]: So just move the code you suggested to a setupFile?

Comment (saitonakamura) [reply src/testUtils/testUtilsThree.tsx]: seems to work just fine

**Pull request merged**